preprocessing/preferences.py

```diff
@@ -0,0 +1,79 @@
+
+import random
+from preprocessing.xml_loader import *
+
+
+def generate_preferences(staff, time_set, num_weekly_preferences, preferences_dur):
+    """
+    Generates random preferences for each employee. The result is normalized to 1
+    Note: the result is identical between different runs, as using the same initial seed will guarantee the same result
+    :param staff: data structure containing employees
+    :param time_set: data structure containing time_step, weeks and needed time_periods-sets
+    :param num_weekly_preferences: (min, max)-tuple representing bounds on the number of requests per employee per week
+    :param preferences_dur: a (min, max)-tuple representing bounds on the durations of the preference
+    :return: a dict with keys [e][t], representing employee and time respectively
+    """
+
+    # Ensuring persistent results between different runs. The seed can be any integer
+    random.seed(1)
+    preferences = tupledict()
+
+    employees = staff["employees"]
+    time_step = time_set["step"]
+    time_periods = time_set["periods"][0]
+    time_periods_in_day = time_set["periods"][2]
+    weeks = time_set["weeks"]
+
+    # Duration in hours is converted to duration in time_steps
+    preferences_dur = [int(preferences_dur[0]/time_step), int(preferences_dur[1]/time_step)]
+
+    # Initializing all preferences
+    for employee in employees:
+        preferences[employee] = tupledict()
+
+        for time in time_periods:
+            preferences[employee][time] = 0
+
+    # Generating preferences
+    for employee in employees:
+        number_of_preferences = 0
+
+        for week in weeks:
+            used_time_periods = []
+            weekly_preferences = int(random.randint(num_weekly_preferences[0], num_weekly_preferences[1]))
+            realized_dur = [random.randint(preferences_dur[0], preferences_dur[1]) for i in range(weekly_preferences)]
+
+            for dur in realized_dur:
+                value = random.choice([-1, 1])
+                unique_preference = False
+
+                # While-block is ensuring that preferences do not overlap each other
+                while not unique_preference:
+                    day = random.randint(week * 7, ((week + 1) * 7) - 1)
+                    unique_preference = True
+                    try:
+                        start_index = random.randint(0, len(time_periods_in_day[day]) - dur)
+                    except:
+                        dur = time_periods_in_day[day][-1] - time_periods_in_day[day][0]
+                        start_index = random.randint(0, len(time_periods_in_day[day]) - dur)
```

Comment:
This is really ugly code. Even though I wrote it myself I am not sure 100% if it will work as intended. Will have to test locally myself.

Reply:
This is reused from last year. The idea, as I see it, is to allow preference intervals that is not long enough. This is vital in cases where demand is only for seven hours, while you are searching for preferences spanning over 8 hours.